cmd/scollector/collectors/aws.go

```diff
@@ -0,0 +1,313 @@
+	descAWSEC2StatusCheck = "The EC2 Status Check, which includes both instance-level and system-level drill-down, gathered every 60 seconds."
+	descAWSELBLatency     = "The minimum, maximum and average latency as reported by the load balancer, gathered at a 60 second interval and averaged over five minutes."
+	descAWSELBHostCount   = "The number of instances in what the Elastic Load Balancer considers a healthy state, gathered every 60 seconds."
+)
+
+func AWS(accessKey, secretKey, region string) {
+	collectors = append(collectors, &IntervalCollector{
+		F: func() (opentsdb.MultiDataPoint, error) {
+			return c_aws(accessKey, secretKey, region)
+		},
+		Interval: 60 * time.Second,
+		name:     fmt.Sprintf("aws-%s", region),
+	})
+}
+
+func c_aws(accessKey, secretKey, region string) (opentsdb.MultiDataPoint, error) {
+	var md opentsdb.MultiDataPoint
+	creds := aws.Creds(accessKey, secretKey, "")
+	if creds == nil {
+		return nil, fmt.Errorf("unable to make creds")
+	}
+	ecc := ec2.New(creds, region, nil)
+	if ecc == nil {
+		return nil, fmt.Errorf("unable to login to EC2")
+	}
+	elb := elb.New(creds, region, nil)
+	if elb == nil {
+		return nil, fmt.Errorf("unable to login to ELB")
+	}
+	cw := cloudwatch.New(creds, region, nil)
+	if cw == nil {
+		return nil, fmt.Errorf("unable to login to CloudWatch")
+	}
+	instances, err := awsGetInstances(*ecc)
+	if err != nil {
+		slog.Info("No EC2 Instances found.")
+	}
+	loadBalancers, err := awsGetLoadBalancers(*elb)
+	if err != nil {
+		slog.Info("No ELB Load Balancecrs found.")
+	}
+	for _, loadBalancer := range loadBalancers {
+		awsGetELBLatency(*cw, &md, loadBalancer)
+		awsGetELBHostCounts(*cw, &md, loadBalancer)
+	}
+	for _, instance := range instances {
+		awsGetCPU(*cw, &md, instance)
+		awsGetNetwork(*cw, &md, instance)
+		awsGetDiskBytes(*cw, &md, instance)
+		awsGetDiskOps(*cw, &md, instance)
+		awsGetStatusChecks(*cw, &md, instance)
+	}
+	return md, nil
+}
+
+func awsGetInstances(ecc ec2.EC2) ([]ec2.Instance, error) {
+	instancelist := []ec2.Instance{}
+	resp, err := ecc.DescribeInstances(nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to describe EC2 Instances")
+	}
+	for _, reservation := range resp.Reservations {
+		for _, instance := range reservation.Instances {
+			instancelist = append(instancelist, instance)
+		}
+	}
+	return instancelist, nil
+}
+
+func awsGetLoadBalancers(lb elb.ELB) ([]elb.LoadBalancerDescription, error) {
+	lbList := []elb.LoadBalancerDescription{}
+	resp, err := lb.DescribeLoadBalancers(nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to describe ELB Balancers")
+	}
+	for _, loadBalancer := range resp.LoadBalancerDescriptions {
+		lbList = append(lbList, loadBalancer)
+	}
+	return lbList, nil
+}
+
+func awsGetCPU(cw cloudwatch.CloudWatch, md *opentsdb.MultiDataPoint, instance ec2.Instance) error {
+	search := cloudwatch.GetMetricStatisticsInput{
```

Comment:
There are many calls to GetMetricStatisticsInput. They each make their own time, which means if one call takes 10 seconds, the results will be skewed with other results by that much. Would it be more useful to instead record now in the main c_aws function, and pass that down to all of these other functions? Then they could compute StartTime based on that, and each function here would be getting data from the same time point.